Add blog post: Using Finalizers to Control Deletion

aaron-kasten opened the pull request:
Add block post describing operation of finalizers and owner references via `kubectl delete`

This blog post contains the content of a presentation given at KubeCon, in blog form.

Submission to `kubenetes.io` recommended by CNCF.

aaron-kasten commented:
@onlydole @sftim email address on git commit is that of our corporate CLA (`aaron@kasten.io` for a CLA for `kasten.io`).

I'm not sure what's missing from the PR to get the CLA detection working correctly.

sftim commented on content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:
````diff
@@ -0,0 +1,270 @@
+Error from server (NotFound): configmaps "mymap" not found
+```
+
+All of the shell commands are in **bold**, and the output follows. You can see that we begin with a `kubectl create configmap mymap`, which will create the empty configmap `mymap`. Next, we need to `get` the configmap to prove it exists. We can then delete that configmap. Attempting to `get` it again produces an HTTP 404 error, which means the configmap is not found.
````
The shell commands are not in bold.

sftim commented on content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:
```diff
@@ -0,0 +1,270 @@
+
+As these examples demonstrate, finalizers can get in the way of deleting resources in Kubernetes, especially when there are parent-child relationships between objects. Often, there is a reason for adding a finalizer into the code, so you should always investigate before manually deleting it. Owner references allow you to specify and remove trees of resources, although finalizers will be honored i ...
+
+All of the scenarios covered in this post are demonstrated in a recent webinar; you can [watch the recording here](https://www.youtube.com/watch?v=F7-ZxWwf4sY).
```
If you want you can embed the video in the page, too. See https://gohugo.io/content-management/shortcodes/#youtube

sftim commented on content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:
```diff
@@ -0,0 +1,270 @@
+
+## Forcing a Deletion of a Namespace
+
+There's one situation that may require forcing finalization for a namespace. If you've deleted a namespace and you've cleaned out all of the objects under it, but the namespace still exists, use the API call `finalize`.This informs the namespace controller that it needs to remove the finalizer from the namespace and perform any cleanup:
```
Is `finalize` an API call or a resource? It looks like a resource (you can `PUT` it) but the resource name is a verb.

aaron-kasten replied:
Changed - hopefully clearer.

Here's a citation to the technique: https://github.com/kubernetes/kubernetes/issues/60807#issuecomment-408599873

sftim commented on content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:
```diff
@@ -0,0 +1,270 @@
+
+Deleting objects in Kubernetes can be challenging. You may think you’ve deleted something, only to find it still persists. While issuing a `kubectl delete` command and hoping for the best might work for day-to-day operations, understanding how Kubernetes `delete` commands operate will help you understand why some objects linger after deletion. 
+
+In this post, we’ll look at:
```
In docs (and documentation-like blog articles), [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we):
```suggestion
In this post, I’ll look at:
```

sftim reviewed (commented):
Hi. “Finaliers” is misspelled, it should be Finalizers.

I've added some other feedback.

Would you be interested in turning this page into a task? For example, “Use finalizers to control deletion” could be an appropriate title.

aaron-kasten commented:
@sftim thank you for the feedback!  I integrated the changes - hopefully an improvement.

aaron-kasten commented:
@sftim what is the next step for approval?  Are there any actions I need to take on this?

sftim commented on content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:
```diff
@@ -0,0 +1,269 @@
+
+As these examples demonstrate, finalizers can get in the way of deleting resources in Kubernetes, especially when there are parent-child relationships between objects. Often, there is a reason for adding a finalizer into the code, so you should always investigate before manually deleting it. Owner references allow you to specify and remove trees of resources, although finalizers will be honored i ...
+
+All of the scenarios covered in this post are demonstrated in this video from a recent webinar: {{< youtube id="F7-ZxWwf4sY" title="Clean Up Your Room! What Does It Mean to Delete Something in K8s">}}
```
For better sizing, try:
```suggestion
{{< youtube youtube-quote-sm id="F7-ZxWwf4sY" title="Clean Up Your Room! What Does It Mean to Delete Something in K8s">}}
```

sftim commented on content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:
```diff
@@ -0,0 +1,269 @@
+
+The state diagram for the basic `delete` command is very simple:
+
+![State diagram for delete](/images/blog/2021-03-03-use-finalizers-to-control-deletion/state-diagram-delete.png)
```
This doesn't render - see https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/03/03/using-finalizers-to-control-deletion/

aaron-kasten replied:
interesting ... I'll reproduce and fix here.

sftim reviewed (changes requested):
A couple more fixes needed I think. Also, would you be willing to squash the commits in this article @aaron-kasten 

(if you're not sure but you're happy to have the changes combined, we can do that squashing for you).

aaron-kasten commented:
@sftim:
* fixed image references (I'll know better for checking image references next time).
* shrunk images a little
* squashed local branch (appears in PR as one commit)
* merged your changes for Youtube link prologue (now gone)

I think that's it... tell me if there's anything else.

aaron-kasten commented:
@sftim how does this PR look.  Is it ready for merge?

sftim commented:
/retitle Add blog post: Using Finalizers to Control Deletion

sftim commented:
Preview: [Using Finalizers to Control Deletion](https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/03/03/using-finalizers-to-control-deletion/)

sftim commented:
We'll follow the [review process](https://github.com/kubernetes/community/tree/master/sig-docs/blog-subproject#review-process) or something roughly like it. At the moment the blog team is a little short on volunteers I'm afraid.

aaron-kasten commented:
@sftim I understand.  Tell me if there's anything I can do to help it along.

sftim commented on content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:
```diff
@@ -0,0 +1,270 @@
+---
+layout: blog
+title: 'Using Finalizers to Control Deletion'
+date: 2021-03-03
```
Can I pencil in a future date here?
```suggestion
date: 2021-04-28
```

aaron-kasten replied:
2021-04-28 works.

aaron-kasten replied:
I used your suggested date.

sftim reviewed (commented):
If you have time, it's tidier to make the filename match the publication date.

I hope we'll get round to a full review soon(ish).

aaron-kasten commented:
> If you have time, it's tidier to make the filename match the publication date.
> 
> I hope we'll get round to a full review soon(ish).

Ugh ... didn't see that - apologies, I'm moving too fast.   I'll change and update.

aaron-kasten commented:
> If you have time, it's tidier to make the filename match the publication date.
> 
> I hope we'll get round to a full review soon(ish).

Looks good now.  Opening for review ...

kbhawkey commented:
/remove-language zh ja ko fr es pt vi id ru pl

kbhawkey commented:
/remove-language de it

sftim commented:
I'll try to review this this week.

sftim commented on content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:
```diff
@@ -0,0 +1,270 @@
+- How the propagation policy can be used to change the order of deletions
+- How deletion works, with examples
+
+For simplicity, all examples will use configmaps and basic shell commands to demonstrate the process. We’ll explore how the commands work and discuss repercussions and results from using them in practice.
```
nit:
```suggestion
For simplicity, all examples will use ConfigMaps and basic shell commands to demonstrate the process. We’ll explore how the commands work and discuss repercussions and results from using them in practice.
```

aaron-kasten replied:
lets go with that!

sftim reviewed (commented):
/lgtm

aaron-kasten commented:
@sftim looks like the suggested change removed the `lgtm` label... I think it needs another `lgtm` (post change).

sftim commented:
/lgtm

rolfedh commented on content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:
```diff
@@ -0,0 +1,270 @@
+
+In short, what’s happened is that the object was updated, not deleted. That’s because Kubernetes saw that the object contained finalizers and put it into a read-only state. The deletion timestamp signals that the object can only be read, with the exception of removing the finalizer key updates. In other words, the deletion will not be complete until we edit the object and remove the finalizer.
+
+Here's a demonstration of using the `patch` command to remove finalizers. If we want to delete and object, we can simply patch it on the command line to remove the finalizers. In this way, the deletion that was running in the background will complete and the object will be deleted. When we attempt to `get` that configmap, it will be gone. 
```
s/and object/an object/

rolfedh commented on content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:
````diff
@@ -0,0 +1,270 @@
+mymap-child   0      13m21s
+```
+
+The --cascade option links to the propagation policy in the API, which allows you to change the order in which objects are deleted within a tree. In the following example uses API access to craft a custom delete API call with a the background propagation policy:
````
s/a the/the/

aaron-kasten commented:
@rolfedh fixups applied.   thank you.

rolfedh commented:
/lgtm

sftim reviewed (commented):
If you can bump the article to say Friday, 2021-04-30, @aaron-kasten

then I think we can probably get this article over the line for that.

Please change both the filename and the `date` in the front matter.

nate-double-u commented:
Hi, I've gone through and tested all the examples in this post verbatim, and they all worked as I expected them to, until I got to the _Forcing a Deletion of a Namespace_ section.

This may be my own inexperience with k8s, but I'm not sure how to set up this example's environment, or see that the snippet did what I was expecting it to do.

Sorry for the long log dump, but I figured it was the best way to show what I saw. I'm using `minikube`, and was trying to correlate outcomes between the CI and the Kubernetes Dashboard.

I had run `kubectl proxy --port=8080 &` earlier in the process 
```
[1] 2817582
Starting to serve on 127.0.0.1:8080
```

Example log leading up to the _Forcing a Deletion of a Namespace_ section:
```
➜  blog-test cat <<EOF | kubectl create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: mymap-parent
EOF
configmap/mymap-parent created

➜  blog-test CM_UID=$(kubectl get configmap mymap-parent -o jsonpath="{.metadata.uid}")

➜  blog-test cat <<EOF | kubectl create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: mymap-child
  ownerReferences:
  - apiVersion: v1
    kind: ConfigMap
    name: mymap-parent
    uid: $CM_UID
EOF
configmap/mymap-child created

➜  blog-test cat <<EOF | curl -X PUT \
  localhost:8080/api/v1/namespaces/test/finalize \
  -H "Content-Type: application/json" \
  --data-binary @-
{
  "kind": "Namespace",
  "apiVersion": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "finalizers": null,
  }
}
EOF
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type Namespace): couldn't get version/kind; json parse error: invalid character '}' looking for beginning of object key string ({\n  \"kind\": \"Namespace\",\n  \"ap ...)",
  "reason": "BadRequest",
  "code": 400
}%                                                                                                                                                       

➜  blog-test curl -X DELETE \
localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
  -d '{
"kind":"DeleteOptions",
"apiVersion":"v1",
"propagationPolicy":"Background"
}' \
  -H "Content-Type: application/json"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Success",
  "details": { ... }
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Success",
  "details": {
    "name": "mymap-parent",
    "kind": "configmaps",
    "uid": "caad82b7-3308-479e-8e8d-b322a25f565b"
  }
}kind:: command not found
apiVersion:: command not found
metadata:: command not found
status:: command not found
details:: command not found

➜  blog-test cat <<EOF | curl -X PUT \
  localhost:8080/api/v1/namespaces/test/finalize \
  -H "Content-Type: application/json" \
  --data-binary @-
{
  "kind": "Namespace",
  "apiVersion": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "finalizers": null,
  }
}
EOF
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type Namespace): couldn't get version/kind; json parse error: invalid character '}' looking for beginning of object key string ({\n  \"kind\": \"Namespace\",\n  \"ap ...)",
  "reason": "BadRequest",
  "code": 400
}% 
```

nate-double-u commented:
(sorry, heads up to the GitHub email notification system users, I've edited the log in my previous post)

sftim commented:
@nate-double-u thanks for taking that time to check!

sftim commented:
```
curl -X DELETE \
localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
  -d '{
"kind":"DeleteOptions",
"apiVersion":"v1",
"propagationPolicy":"Background"
}' \
  -H "Content-Type: application/json"
```

The problem here is wrapping I think.

```bash
curl -X DELETE \
  http://localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
  -d '{ "kind":"DeleteOptions", "apiVersion":"v1", "propagationPolicy":"Background" }' \
  -H "Content-Type: application/json"
```
ought to work fine

nate-double-u commented:
> ```shell
> curl -X DELETE \
>   http://localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
>   -d '{ "kind":"DeleteOptions", "apiVersion":"v1", "propagationPolicy":"Background" }' \
>   -H "Content-Type: application/json"
> ```
> 
> ought to work fine

Confirmed, this works. Thanks!

nate-double-u commented on content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:
````diff
@@ -0,0 +1,270 @@
+
+There's one situation that may require forcing finalization for a namespace. If you've deleted a namespace and you've cleaned out all of the objects under it, but the namespace still exists, deletion can be forced by updating the namespace subresource, `finalize`. This informs the namespace controller that it needs to remove the finalizer from the namespace and perform any cleanup:
+
+```
+$ cat <<EOF | curl -X PUT \
+  localhost:8080/api/v1/namespaces/test/finalize \
+  -H "Content-Type: application/json" \
+  --data-binary @-
+{
+  "kind": "Namespace",
+  "apiVersion": "v1",
+  "metadata": {
+    "name": "test"
+  },
+  "spec": {
+    "finalizers": null,
+  }
+}
+EOF
+```
````
Please see https://github.com/kubernetes/website/pull/26997#issuecomment-829063409

aaron-kasten commented:
> ```shell
> ```shell
> curl -X DELETE \
>   http://localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
>   -d '{ "kind":"DeleteOptions", "apiVersion":"v1", "propagationPolicy":"Background" }' \
>   -H "Content-Type: application/json"
> ```
> 
> 
> ought to work fine
> 
> Confirmed, this works. Thanks!
> ```

@nate-double-u is there action I should take on this?  This sounds like a layout issue?

nate-double-u commented:
> @nate-double-u is there action I should take on this? This sounds like a layout issue?

I think that if folks are copying and pasting snippets from the article, then @sftim's layout works better (though it may be less readable as several bits of info are put onto the same line).

I think I'd err on the side of things working for the most readers when copied from the article, so I'd suggest updating the layout. That said, this may be an issue with my setup (using zsh instead of bash), so I'd defer the decision to you others who might have an opinion (does it work as is for other users?)

aaron-kasten commented:
> > @nate-double-u is there action I should take on this? This sounds like a layout issue?
> 
> I think that if folks are copying and pasting snippets from the article, then @sftim's layout works better (though it may be less readable as several bits of info are put onto the same line).
> 
> I think I'd err on the side of things working for the most readers when copied from the article, so I'd suggest updating the layout. That said, this may be an issue with my setup (using zsh instead of bash), so I'd defer the decision to you and others who might have an opinion (does it work as is for other users?)

@nate-double-u agreed - I think a least-common-denominator approach is best, which is represented by keeping everything on one line.  The JSON is not so large that it makes the example confusing.

aaron-kasten commented:
@sftim @nate-double-u changes are in ... all set.

nate-double-u commented:
new date deploy preview: [Using Finalizers to Control Deletion](https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/04/30/using-finalizers-to-control-deletion/)

nate-double-u commented on content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md:
````diff
@@ -0,0 +1,267 @@
+
+There's one situation that may require forcing finalization for a namespace. If you've deleted a namespace and you've cleaned out all of the objects under it, but the namespace still exists, deletion can be forced by updating the namespace subresource, `finalize`. This informs the namespace controller that it needs to remove the finalizer from the namespace and perform any cleanup:
+
+```
+$ cat <<EOF | curl -X PUT \
+  localhost:8080/api/v1/namespaces/test/finalize \
+  -H "Content-Type: application/json" \
+  --data-binary @-
+{
+  "kind": "Namespace",
+  "apiVersion": "v1",
+  "metadata": {
+    "name": "test"
+  },
+  "spec": {
+    "finalizers": null,
+  }
+}
+EOF
+```
+
````
I'm still having trouble getting this snippet to work. I'm not sure if the setup is intended to be the same as the examples above, but I always get the following:

```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type Namespace): couldn't get version/kind; json parse error: invalid character '}' looking for beginning of object key string ({\n  \"kind\": \"Namespace\",\n  \"ap ...)",
  "reason": "BadRequest",
  "code": 400
}% 
```

aaron-kasten replied:
there was a dangling `,` - removed it.  Example should work now.

aaron-kasten replied:
@nate-double-u done.

kbhawkey commented on content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md:
````diff
@@ -0,0 +1,267 @@
+
+```
+$ kubectl create configmap mymap
+configmap/mymap created
````
minor nit: You could remove the shell prompt ($)

aaron-kasten replied:
I removed the command prompts.

aaron-kasten replied:
@kbhawkey done.

nate-double-u reviewed (commented):
I'm still not sure how to get k8s into a state described in the _Forcing a Deletion of a Namespace_ section, but when I create a new test namespace (`kubectl create namespace test`) first, the snippet doesn't throw the `BadRequest` error.

Thanks for this @aaron-kasten, it's an interesting article!

LGTM

aaron-kasten commented:
@sftim @kbhawkey unsure of what the next steps are.  The PR still needs a `lgtm` from a reviewer.

aaron-kasten commented:
@sftim is this ready?

sftim commented:
I've set a publication date for Friday
/label tide/merge-method-squash
